plugins/woocommerce/includes/class-wc-cart-session.php

```diff
@@ -282,12 +282,14 @@ public function get_cart_from_session() {
 		$cart_for_session = $this->get_cart_for_session();
 
 		if ( empty( $cart_for_session ) ) {
-			WC()->session->set( 'cart', null );
-			$update_cart_session = true;
-		}
-
-		if ( $update_cart_session ) {
+			// If the cart is empty, clear the cart session directly.
+			$this->destroy_cart_session();
+		} elseif ( $update_cart_session ) {
+			// If the cart is not empty, and the cart session needs to be updated, calculate totals. Session will update after this.
 			$this->cart->calculate_totals();
+		} else {
+			// Otherwise, just set the session. This was previously hooked into `woocommerce_cart_loaded_from_session` but that resulted in multiple session updates.
+			$this->set_session();
```

Comment:
This prevents multiple calls to `set_session`. If we have no cart, we can destroy the session with `destroy_cart_session` without triggering a cart recalculation. If we're updating the session, we can call calculate_totals; set_session is hooked into that when complete. Otherwise, just set the session directly. No need to use `woocommerce_cart_loaded_from_session` hook here.